toolchain/check/impl_lookup.cpp

```diff
@@ -561,35 +537,73 @@ auto DeferredLookupSingleImplWitness(Context& context, SemIR::LocId loc_id,
+  SemIR::ConstantId query_self_const_id =
+      context.constant_values().Get(eval_query.query_self_inst_id);
+  SemIR::SpecificInterfaceId query_specific_interface_id =
+      eval_query.query_specific_interface_id;
+
+  // NOTE: Do not retain this reference to the SpecificInterface obtained from a
+  // value store by SpecificInterfaceId. Doing impl lookup does deduce which can
+  // do more impl lookups, and impl lookup can add a new SpecificInterface to
+  // the store which can reallocate and invalidate any references held here into
+  // the store.
+  auto query_specific_interface =
+      context.specific_interfaces().Get(query_specific_interface_id);
+
+  // When the query is a concrete FacetValue, we want to look through it at the
+  // underlying type to find all interfaces it implements. This supports
+  // conversion from a FacetValue to any other possible FacetValue, since
+  // conversion depends on impl lookup to verify it is a valid type change. See
+  // https://github.com/carbon-language/carbon-lang/issues/5137. We can't do
+  // this step earlier than inside impl lookup since:
+  // - We want the converted facet value to be preserved in
+  //   `FindWitnessInFacet()` to avoid looking for impl declarations.
+  // - The constant self value may be modified during constant evaluation as a
+  //   more specific value is found.
+  if (auto facet_value = context.insts().TryGetAs<SemIR::FacetValue>(
+          context.constant_values().GetInstId(query_self_const_id))) {
+    query_self_const_id =
+        context.constant_values().Get(facet_value->type_inst_id);
+  }
+
+  auto query_type_structure = BuildTypeStructure(
+      context, context.constant_values().GetInstId(query_self_const_id),
+      query_specific_interface);
+
+  auto candidate_impls = CollectCandidateImplsForQuery(
+      context, query_type_structure, query_specific_interface);
+
+  for (const auto& candidate : candidate_impls) {
+    // In deferred lookup for a symbolic impl witness, while building a
+    // specific, there may be no stack yet as this may be the first lookup. If
+    // further lookups are started as a result in deduce, they will build the
+    // stack.
+    //
+    // NOTE: Don't retain a reference into the stack, it may be invalidated if
+    // we do further impl lookups when GetWitnessIdForImpl() does deduction.
+    if (!context.impl_lookup_stack().empty()) {
+      context.impl_lookup_stack().back().impl_loc = candidate.loc_inst_id;
     }
+
     // NOTE: GetWitnessIdForImpl() does deduction, which can cause new impls
     // to be imported, invalidating any pointer into `context.impls()`.
     auto result = GetWitnessIdForImpl(
-        context, loc_id, query_is_concrete, query_self_const_id,
-        query_specific_interface, candidate.impl_id);
-    if (std::holds_alternative<UseOriginalImplSymbolicWitness>(result)) {
+        context, loc_id,
+        QueryIsConcrete(context, query_self_const_id, query_specific_interface),
```

Comment:
Can this call be moved back outside the loop?

Reply:
Oh right, yes.